src/main/java/org/assertj/core/condition/MappedCondition.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * Copyright 2012-2021 the original author or authors.
+ */
+package org.assertj.core.condition;
+
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import org.assertj.core.annotations.Beta;
+import org.assertj.core.api.Condition;
+
+/**
+ * Container-<code>{@link Condition}</code> that does a mapping and then uses nested
+ * <code>{@link Condition}</code> to test the mapped actual value.
+ *
+ * <pre><code class='java'> Condition&lt;String&gt; hasLineSeparator = new Condition&lt;&gt;(t -&gt; t.contains(System.lineSeparator()), "has lineSeparator");
+ *
+ * Condition&lt;Optional&lt;String&gt;&gt; optionalHasLineSeparator = MappedCondition.mappedCondition(Optional::get, hasLineSeparator);
+ *
+ * // returns true
+ * optionalHasLineSeparator.matches(Optional.of("a" + System.lineSeparator()));
+ *
+ * // returns false
+ * optionalHasLineSeparator.matches(Optional.of("a"));</code></pre>
+ *
+ * @param <FROM> the type of object this condition accepts.
+ * @param <TO> the type of object the nested condition accepts.
+ *
+ * @author Stefan Bischof
+ */
+@Beta
+public class MappedCondition<FROM, TO> extends Condition<FROM> {
+
+  private Condition<TO> condition;
+  private Function<FROM, TO> mapping;
+  private String mappingDescription;
+
+  /**
+   * Creates a new <code>{@link MappedCondition}</code>
+   *
+   * @param <FROM> the type of object the given condition accept.
+   * @param <TO> the type of object the nested condition accept.
+   * @param mapping the Function that maps the value to test to the a value for the nested condition.
+   * @param mappingDescription describes the mapping verbal.
+   * @param condition the nested condition to evaluate.
+   * @return the created {@code Mapped}.
+   * @throws NullPointerException if the given condition is {@code null}.
+   * @throws NullPointerException if the given mapping is {@code null}.
+   */
+  public static <FROM, TO> MappedCondition<FROM, TO> mappedCondition(Function<FROM, TO> mapping, String mappingDescription,
+                                                                     Condition<TO> condition) {
```

Comment:
@joel-costigliola just curious if there is any reason **mappingDescription** is not last argument? That way a varargs method can be introduced like in Condition class.

Reply:
good idea!